File: java/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderApprovalUseCase.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.repository.OrderRepository;

public class OrderApprovalUseCase {
    private final OrderRepository orderRepository;

    public OrderApprovalUseCase(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public void run(OrderApprovalRequest request) {
        final Order order = orderRepository.getById(request.getOrderId());

//        order.validateRequest(request)
        if(request.isApproved()) {
            order.approve();
        } else {
            order.reject();
        }
        orderRepository.save(order);
    }
}

File: kotlin/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderApprovalUseCase.kt
package it.gabrieletondi.telldontaskkata.useCase

import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.repository.OrderRepository

class OrderApprovalUseCase(private val orderRepository: OrderRepository) {
    fun run(request: OrderApprovalRequest) {
        val order = orderRepository.getById(request.orderId)
        if (order!!.status == OrderStatus.SHIPPED) {
            throw ShippedOrdersCannotBeChangedException()
        }
        if (request.isApproved && order.status == OrderStatus.REJECTED) {
            throw RejectedOrderCannotBeApprovedException()
        }
        if (!request.isApproved && order.status == OrderStatus.APPROVED) {
            throw ApprovedOrderCannotBeRejectedException()
        }
        order.status = if (request.isApproved) OrderStatus.APPROVED else OrderStatus.REJECTED
        orderRepository.save(order)
    }
}